### ISSUE TITLE
[$500] HIGH: Review who can remove whom from groups, workspaces, and workspace objects

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#33575]
**Problem:**
There are complex rules for which room members can see other room members -- rules that were recently updated because there wasn't widespread agreement on how it should work, and how it did work wasn't really to anyone's liking.  There's a risk we're doing the same with who can remove people from the room.

**Solution:**
As discussed [in Slack](https://expensify.slack.com/archives/C0671JDRKQW/p1703547218740349?thread_ts=1703547187.140519&cid=C0671JDRKQW), these are what the rules *should* be, as documented in [this SO](https://stackoverflowteams.com/c/expensify/questions/18263/18264#18264):

* **Nobody can leave or be removed from something they were automatically added to.**  For example:
    - [ ] DM members can't leave or be removed from their DMs
    - [ ] Members can't leave or be removed from their own workspace chats
    - [ ] Admins can't leave or be removed from workspace chats
    - [ ] Members can't leave or be removed from the #announce room
    - [ ] Admins can't leave or be removed from #admins
    - [ ] Domain members can't leave or be removed from their domain chat
    - [ ] Report submitters can't leave or be removed from their reports (eg, if they are the `report.accountID`)
    - [ ] Report managers can't leave or be removed from their reports (eg, if they are the `report.managerID`)
    - [X] ~Group owners cannot be removed from their groups -- they need to transfer ownership first~
        - Note: Groups don't exist yet, so you can skip this 

* **Excepting the above, admins can remove anyone.** For example:
    - [X] ~Group admins can remove other group admins, as well as group members~
    - [ ] Workspace admins can remove other workspace admins, as well as workspace members, and invited guests

* **Excepting the above, members can remove guests.** For example:
    - [ ] Workspace members can remove non-workspace guests.

* **Excepting the above, anybody can remove themselves from any object**
    - [ ] Confirmed

As part of this GH please do the following:

1. Document the above rules in https://github.com/Expensify/App/blob/main/README.md in a new section called **Security**, put between **Philosophy** and **Internationalization** (add to the ToC and copy the header styles of the surrounding sections)

2. Find every place in the app that involves adding/removing people to an object's membership, and add a link to that new security section in a comment (eg, `// Please see https://github.com/Expensify/App/blob/main/README.md#Security for more details`)

3. Verify that each of the above rules is being correctly enforced by:
    1. Create the conditions to test the rule (eg, for "DM members can't leave or be removed from their DMs", open a test DM to another user)
    2. Manually trigger an API command that attempts to do the *wrong* thing (eg, attempt to remove either of the DM participants) -- ideally the server will correctly reject it, but if not, record it in this issue so we can fix the server.
    3. Verify that there is no method in the client-side app to attempt the forbidden action (eg, in the case of the DM there is no UI to show the member list, so therefore the user can't attempt it)
    4. Add a comment to this GH capturing evidence that you did this test (ie, a screenshot or console logs)
    5. Check off the test in the OP and go to the next

4. When all the boxes are checked, submit the PR or the new Security section and all the corresponding comment links.

---

### Proposals work differently on this issue

- For all the checkboxes above, please identify what features are present in the App and can be worked right now. For features missing in the App like "Leaving workspace chat or #admins room or #announce room". Please flag them so we can discuss and create an issue for them
- For each checkbox, please briefly explain where and why you want to make the change to fix the access control
- Addition of jest tests is a plus!

As mentioned by David here https://github.com/Expensify/App/issues/33575#issuecomment-1872623804
> Great question, this is a kind of different issue than normally. So yes, we need some way to choose who to do this, and "first come first serve" doesn't seem very discerning. Can you make a proposal that explains why we should pick you over someone else? I'm not sure what that would entail -- maybe cite past experience, or show that you've already worked in this area of the code, or even pick one of the more complex ones to talk about how you would go about testing it? Something to stand out.

We'll choose the proposal that nearly fixes this issue and help us to find and figure out missing pieces

Thanks!
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fb7403bad51be652</li>
        <li>Upwork Job ID: 1741227152068182016</li>
        <li>Last Price Increase: 2024-01-13</li>
<li>Automatic offers: </li>
<ul>
<li>aimane-chnaif | Reviewer | 28105369</li>
</ul></li>
    </ul>
</details>

**Comment by abeebridwan** [comment]
interested

**Comment by quinthar** [comment]
Great question, this is a kind of different issue than normally.  So yes, we need *some* way to choose who to do this, and "first come first serve" doesn't seem very discerning.  Can you make a proposal that explains why we should pick you over someone else?  I'm not sure what that would entail -- maybe cite past experience, or show that you've already worked in this area of the code, or even pick one of the more complex ones to talk about how you would go about testing it?  Something to stand out.

**Comment by quinthar** [comment]
@MonilBhavsar what are the next steps?  Do you select a proposal?  Or should this have been assigned a C+?

**Comment by MonilBhavsar** [comment]
As per the process, I would like @aimane-chnaif to select a proposal and then I can finally assign them to the issue, if that sounds good

**Comment by MonilBhavsar** [comment]
I have updated the github issue description on posting proposals for this issue. cc @quinthar @aimane-chnaif 
@rayane-djouah you might want to update the proposal based on that update

**Comment by MonilBhavsar** [comment]
> Currently, the app does not provide features that attempt the forbidden actions. we should normalize this so developers would not attempt to include this features in the future. and we should ensure that the server reject these actions if someone attempts to do it by an API request.

Could you please elaborate what do you mean by - "we should normalize this so developers would not attempt to include this features in the future". 

> Internal team will need to add automated tests for this rules to the backend code in the future.

We'll definitely have server side tests and manual regression steps, but having jest tests to check disabled button or alert if not permitted action is also preferred

**Comment by quinthar** [comment]
What's the next step here?  Who are we blocked on why?

**Comment by MonilBhavsar** [comment]
Have asked for help here https://expensify.slack.com/archives/C03UK30EA1Z/p1705403120723279

**Comment by waterim** [comment]
Hello, I'm Artem from Callstack - expert group.
And I would like to help with this issue

**Comment by quinthar** [comment]
Great!  @waterim can you please outline your proposal here?

**Comment by waterim** [comment]
**Nobody can leave or be removed from something they were automatically added to. For example:**

- [X]  DM members can't leave or be removed from their DMs
- [X]  Members can't leave or be removed from their own workspace chats
- [X]  Admins can't leave or be removed from workspace chats
- [X]  Members can't leave or be removed from the #announce room
- [X]  Admins can't leave or be removed from #admins
- [X]  Domain members can't leave or be removed from their domain chat?
- [X]  Report submitters can't leave or be removed from their reports (eg, if they are the report.accountID)
- [X]  Report managers can't leave or be removed from their reports (eg, if they are the report.managerID)

**Excepting the above, admins can remove anyone. For example:**

- [X]  Workspace admins can remove other workspace admins, as well as workspace members, and invited guests

**Excepting the above, members can remove guests. For example:**

- [X]  Workspace members can remove non-workspace guests.
A little bit weird behaviour that invited workspace member can remove creator of the room. I think it can be redesigned that creator of the room can't be deleted by someone who was added there.

**Excepting the above, anybody can remove themselves from any object**

- [X]  Confirmed


Looks like all of those cases works fine, what we need to do is to:
1. Document all those rules.
2. Add comments to each of the part of the code where deleting is happening with "RemoveFromRoom" API call.
3. Write all tests and test a backend for all those checkboxes above.

**Comment by MonilBhavsar** [comment]
@waterim currently members can't leave or be removed from #admins, #announce or workspace chats i.e. the functionality doesn't exist, right?
I think we should create issue to add those features, and make it follow the rules mentioned in this issue

**Comment by waterim** [comment]
@MonilBhavsar yes, it doesn’t exist, the only way is to remove user from the workspace fully.
Do we have some information like who can leave/remove people from workspace chats?

**Comment by quinthar** [comment]
> Do we have some information like who can leave/remove people from workspace chats?

Applying the above rules, I think this means:

1. Each member is added automatically to their own workspace chat, which means members cannot be removed from their own worksapce chats
2. Each admin is added automatically to all workspace chats, which means that admins cannot be removed from any workspace chat
3. Anybody else can be freely added to a workspace chat via invite or mention, meaning they can also be freely removed either by themselves, or by admins

Does this sound right?

**Comment by MonilBhavsar** [comment]
Agree, sounds good to me 👍

**Comment by quinthar** [comment]
@waterim Do you have an ETA for the comment and doc updates?

**Comment by MonilBhavsar** [comment]
@waterim I see a PR linked to the issue. Nice!
Could you please tell what you'll be implementing in PR.

And, what do you think of the comment here https://github.com/Expensify/App/issues/33575#issuecomment-1902784809
Are we creating a new issue to handle leaving/removing members from workspace chats?

**Comment by waterim** [comment]
Hey @MonilBhavsar In PR I already added Security section and links, I think for this PR thats will be it.
To be clear:
"3. Verify that each of the above rules is being correctly enforced by:" - for this part I need to write test steps for all the rules and test it, also test backend API calls, right?

If yes, the only thing left is to test everything fully.

"Anybody else can be freely added to a workspace chat via invite or mention, meaning they can also be freely removed either by themselves, or by admins

From the https://github.com/Expensify/App/issues/33575#issuecomment-1902784809 the only 3d point is 
> 3. Anybody else can be freely added to a workspace chat via invite or mention, meaning they can also be freely removed either by themselves, or by admins

"meaning they can also be freely removed either by themselves, or by admins" This part here looks like is not implemented right now. Because its not possible to leave from the workspace.

Yes, I think for this we need to create a new issue for leaving/removing members from workspace chats.
Remove already exists, admins can freely remove users, but its not possible to leave.
We need to create a ticket "Add possibility to leave from workspace"

**Comment by waterim** [comment]
@MonilBhavsar Have a question:
"Members can't leave or be removed from their own workspace chats"
Does it mean that members can't leave or be removed from a workspace they created or it means that members cant be removed from a "workspace room" they created?

**Comment by waterim** [comment]
# Conditions to test rules: 
### DM members can't leave or be removed from their DMs
1. Open/Create any DM with another user
2. Check that no "Remove/Add/Leave" button exists
<img width="1359" alt="image" src="https://github.com/Expensify/App/assets/39777589/dc611610-4557-42c3-abc5-3cf95ae82101">

### Members can't leave or be removed from their own workspace chats
1. Create a workspace/Go to any workspace you created
2. Add member/admin to this workspace
3. Login as this newly added user.
4. Go to workspaces
5. Select your workspace
6. Go to Members
7. Try to remove creator
<img width="382" alt="image" src="https://github.com/Expensify/App/assets/39777589/81184072-42d9-41ca-b7b3-f5e60f0a98fe">


### Admins can't leave or be removed from workspace chats
1. Create a workspace/Go to any workspace you created
2. Add member/admin to this workspace
3. Login as this newly added user.
4. Go to workspaces
5. Select your workspace
6. Go to Members
7. Try to remove creator
<img width="382" alt="image" src="https://github.com/Expensify/App/assets/39777589/81184072-42d9-41ca-b7b3-f5e60f0a98fe">

### Members can't leave or be removed from the #announce room
1. Go to #announce room 
2. Try to remove user
3. Look that no "remove" button on the screen
<img width="1347" alt="image" src="https://github.com/Expensify/App/assets/39777589/a4eefd45-bc17-4fb7-a466-97bc951b79a3">


### Admins can't leave or be removed from #admins
1. Go to #admin room 
2. Try to remove user
3. Look that no "remove" button is on the screen
<img width="1369" alt="image" src="https://github.com/Expensify/App/assets/39777589/7ddd97db-08d4-45de-92d6-919820ab2f18">


### Domain members can't leave or be removed from their domain chat
### Report submitters can't leave or be removed from their reports
1. Create a request from a workspace with someone in "Submit and Approve role"
2. Try to remove user
3. Look that no "remove" button is on the screen
<img width="1359" alt="image" src="https://github.com/Expensify/App/assets/39777589/56411cb3-715a-4394-a677-9986d4ab0e1c">


### Report managers can't leave or be removed from their reports
1. Create a request from a workspace with someone in "Submit and Approve role"
2. Try to remove user
3. Look that no "remove" button is on the screen
<img width="1359" alt="image" src="https://github.com/Expensify/App/assets/39777589/56411cb3-715a-4394-a677-9986d4ab0e1c">

### Workspace admins can remove other workspace admins, as well as workspace members, and invited guests
1. Go to workspace settings
2. Go to members
3. Try to remove admin/member/guest
4. There is a possibility to remove all of them 
<img width="416" alt="image" src="https://github.com/Expensify/App/assets/39777589/5f0920ec-1604-4ee7-81dd-a7dda2943481">

### Workspace members can remove non-workspace guests.
1. Create a room within a workspace
2. Invite someone from the workspace and someone from the outside
3. Login as a member of the workspace, but not as creator of room
4. Try to delete a guest
<img width="496" alt="image" src="https://github.com/Expensify/App/assets/39777589/9d0098ed-f262-4727-954f-dfd6cae61710">

But its also a little bit weird behaviuor, because Its possible to remove a room creator(artem.makushov@callstack.com)
@MonilBhavsar @quinthar - Is it a correct behaviour or we need to create a new issue for that?

**Comment by quinthar** [comment]
>"3. Verify that each of the above rules is being correctly enforced by:" - for this part I need to write test steps for all the rules and test it, also test backend API calls, right?

Yes

**Comment by quinthar** [comment]
> "meaning they can also be freely removed either by themselves, or by admins" This part here looks like is not implemented right now. Because its not possible to leave from the workspace.
>
>Yes, I think for this we need to create a new issue for leaving/removing members from workspace chats.
> Remove already exists, admins can freely remove users, but its not possible to leave.
> We need to create a ticket "Add possibility to leave from workspace"

Great!  Can you create a new issue with detailed instructions on what to do, and then link the bullet in the original post to the new issue, so we can keep track of what needs to be done?

**Comment by quinthar** [comment]
> But its also a little bit weird behaviuor, because Its possible to remove a room creator([artem.makushov@callstack.com](https://www.expensify.com/_support/?supportEmail=artem.makushov%40callstack.com&comment=better%20supportal%20issue%20nubmer%3A%2033575&location=_support/))
> @MonilBhavsar @quinthar - Is it a correct behaviour or we need to create a new issue for that?

Hm.... I think it's fine to remove a room creator (and it's fine for them to leave).

**Comment by quinthar** [comment]
Just confirming, are you following these steps in order?  If so, what number are you on?
<img width="826" alt="image" src="https://github.com/Expensify/App/assets/575186/26f14923-039a-4773-836a-2883082b3b58">

**Comment by quinthar** [comment]
Earlier you mentioned EOW as a possible ETA, but it kinda seems unlikely that you'll finish by tomorrow, given that none of the boxes are checked.   Can you give an updated ETA?

**Comment by waterim** [comment]
@quinthar yes, Im following these steps, already added security section and links into the draft PR(was waiting for answers to continue on that)
Conditions to test those rules you can see as my previous message. 

Just backend tests, create new issue and few improvements on the PR left for this issue


Regarding the EOW I think I was too optimistic, because didn’t take into account a time of the responses, but will try to create a new issue, test the backend and adjust security section today, new estimated ETA is a Tuesday/Wednesday because Im sure changes to security section and some other stuff will require improvements from PR reviewer and you.

**Comment by quinthar** [comment]
Great!

**Comment by waterim** [comment]
@quinthar Created an issue, but I don't have access to change anything except of the text on the ticket.
Can't add labels, main project etc.

**Comment by waterim** [comment]
> 3\. Anybody else can be freely added to a workspace chat via invite or mention, meaning they can also be freely removed either by themselves, or by admins

@quinthar @MonilBhavsar 
Regarding this statement:
Workspace chat can have only user and all admins, but Its not possible to add any other members/guests to those chats.
Am I right? Because I tested this statement and it just doesnt work.

If Im right do we need to create a ticket to add this feature?

**Comment by MonilBhavsar** [comment]
Catching up on this. Everything looks good 👍 

> > But its also a little bit weird behaviuor, because Its possible to remove a room creator([artem.makushov@callstack.com](https://www.expensify.com/_support/?supportEmail=artem.makushov%40callstack.com&comment=better%20supportal%20issue%20nubmer%3A%2033575&location=_support/))
> > @MonilBhavsar @quinthar - Is it a correct behaviour or we need to create a new issue for that?
> 
> Hm.... I think it's fine to remove a room creator (and it's fine for them to leave).

Just to clarify here - I think we need atleast one room admin in the room to manage the room. If the room creator is the only admin and they leave the room, then there would be no admin to manage the room

**Comment by MonilBhavsar** [comment]
> > 3. Anybody else can be freely added to a workspace chat via invite or mention, meaning they can also be freely removed either by themselves, or by admins
> 
> @quinthar @MonilBhavsar Regarding this statement: Workspace chat can have only user and all admins, but Its not possible to add any other members/guests to those chats. Am I right? Because I tested this statement and it just doesnt work.
> 
> If Im right do we need to create a ticket to add this feature?

That sounds correct to me 👍 We would need to add that feature

**Comment by waterim** [comment]
How it should look like?
Just by mention we are adding members and then this member should be able to leave or be removed from this chat, right?
@quinthar @MonilBhavsar 

If yes - I can create a new issue for that.

**Comment by waterim** [comment]
Also tested backend for those rules, no forbidden actions!
Updated with screenshots with responses [here](https://github.com/Expensify/App/issues/33575#issuecomment-1908260054)

**Comment by quinthar** [comment]
>Also I think there should be a separate ticket for "Leave from the workspace", or we can combine it together like: "Add possibility to leave from the workspace and workspace chats"

This issue is great.  Leaving from the workspace will automatically make you leave from the workspace chat.

**Comment by quinthar** [comment]
> Workspace chat can have only user and all admins, but Its not possible to add any other members/guests to those chats. Am I right? Because I tested this statement and it just doesnt work.

This is not correct.  Anybody can be invited to any chat, the sole exception being DM chats, which only ever have 2 participants.  But literally anybody can be added to literally any other chat.   So I can mention you into my workspace chat, and it'll work fine.

**Comment by quinthar** [comment]
> Just to clarify here - I think we need atleast one room admin in the room to manage the room. If the room creator is the only admin and they leave the room, then there would be no admin to manage the room
There's no such thing as a "room admin".  I think it's fine to remove the creator, as the creator gets no special treatment -- nothing breaks if they are removed, so far as I know.

**Comment by quinthar** [comment]
> Just by mention we are adding members and then this member should be able to leave or be removed from this chat, right?
I don't quite understand this question.  Can you elaborate on it?

**Comment by waterim** [comment]
> This is not correct. Anybody can be invited to any chat, the sole exception being DM chats, which only ever have 2 participants. But literally anybody can be added to literally any other chat. So I should be able to mention you into my workspace chat, and it should work fine.

Okay, then I will create a new ticket for that. Because it doesn’t work now, I tried several times in workspace chat to mention different users(workspace members and not) and its not inviting this user

**Comment by waterim** [comment]
> > Just by mention we are adding members and then this member should be able to leave or be removed from this chat, right?
> 
> I don't quite understand this question. Can you elaborate on it?

As you said we should be able to invite new members to the workspace chats by „mentioning them”.
Admin and default member of the workspace chat are not able to leave or be removed from this workspace chat, but in the same time this „invited” user should be able to leave and admin/default member should be able to remove him, right?

**Comment by quinthar** [comment]
Aha, great catch, please make an issue, thanks!

**Comment by quinthar** [comment]
> Okay, then I will create a new ticket for [mentioning users to add them to workspace chats]. Because it doesn’t work now, I tried several times in workspace chat to mention different users(workspace members and not) and its not inviting this user

Yes please do, thanks!

**Comment by quinthar** [comment]
> As you said we should be able to invite new members to the workspace chats by „mentioning them”.
Admin and default member of the workspace chat are not able to leave or be removed from this workspace chat, but in the same time this „invited” user should be able to leave and admin/default member should be able to remove him, right?

Correct.

**Comment by quinthar** [comment]
Still on track for tomorrow or Wed?

**Comment by waterim** [comment]
@quinthar if everything is fine with PR - yes, we can finish tomorrow.
I opened a PR day before yesterday, but no reviews for now! Also yesterday updated a security section and tested manually backend. 

Will create an issue now and waiting for the reviews.

Also I will be OOO tomorrow and on Thursday(2 days) because of the sick leave and if something urgent will appear @rezkiy37 can help with finishing it before I come back!

**Comment by waterim** [comment]
New issues created from this ticket:
1. Add possibility to leave from workspace
2. Add users to workspace chat by mentioning them.
3. Add possibility to remove invited user from workspace chat and add possibility for invited user to leave from the workspace chat

**Comment by MonilBhavsar** [comment]
Sounds good, thanks!

**Comment by quinthar** [comment]
What remains to finish this?  I saw "I opened a PR day before yesterday, but no reviews for now" who is this waiting on?  What do they need to do, and when will it be done?

**Comment by MonilBhavsar** [comment]
@aimane-chnaif yes PR is ready for review! 
For testing can we please confirm that the rules we added in README.md are working as expected in the App. I'll update Tests section accordingly.


Also, Reviewing label was removed because that is not the only PR required for this issue or that PR does not completely fixes/closes this issue 😄

**Comment by MonilBhavsar** [comment]
PR to update Security settings in README is merged

**Comment by MonilBhavsar** [comment]
Going to close this issue now given we have two children issues https://github.com/Expensify/App/issues/35391 and  https://github.com/Expensify/App/issues/35286 in vip-vsb and wave-8 respectively
